### PR TITLE
add sasl SCRAM-SHA-256 authentication mechanism

### PR DESCRIPTION
**Pull request by benbenw**
support SASL SCRAM-SHA-256 as an authentication mechanism.

SCRAM-SHA-256-PLUS (added in Postgresql 11) is **not** supported.

The scram part is provided by _om.ongres.scram:client_ (BSD 2-Clause "Simplified" License), this is a new dependency.
This library also provides the scram support in pgjdbc.

Please review carefully, I'm not intimate with the vertx-pg-client source code.

### CODE REVIEW

**Comment by vietj** [comment]
can you sign the ECA @benbenw ?

**Comment by benbenw** [comment]
@vietj done

**Comment by benbenw** [comment]
@vietj just added a test based on org.testcontainers (already used in the mysql driver)

the sals scram connection is now tested with the test suite

**Comment by benbenw** [comment]
travis test failed on postgresql 9

sasl scram is available starting with postgresql 10

will ignore the test if postgresql < v10

**Comment by benbenw** [comment]
@vietj do you anything more for this PR ?

File: pom.xml [hunk 1]
```diff
@@ -59,6 +59,7 @@
     <!-- We skip sources jar generation as we do it with the assembly plugin to have greater control over the content -->
     <source.skip>true</source.skip>
     <kotlin.version>1.3.0</kotlin.version>
+    <testcontainers.version>1.12.1</testcontainers.version>
```
Comment (BillyYccc) [inline pom.xml]: move this to `vertx-pg-client`

File: vertx-mysql-client/pom.xml [hunk 2]
```diff
@@ -67,6 +67,12 @@
       <version>1.12.0</version>
       <scope>test</scope>
     </dependency>
+    <dependency>
+			<groupId>org.testcontainers</groupId>
+			<artifactId>postgresql</artifactId>
+			<version>1.12.1</version>
+			<scope>test</scope>
+		</dependency>
```
Comment (BillyYccc) [inline vertx-mysql-client/pom.xml]: why is this needed?

File: vertx-pg-client/src/test/java/io/vertx/pgclient/junit/ContainerPgRule.java [hunk 3]
```diff
@@ -0,0 +1,162 @@
...
+    }
+  }
+
+  @Override
+  public Statement apply(Statement base, Description description) {
+    return new Statement() {
+      @Override
+      public void evaluate() throws Throwable {
+        if (!isAtLeastPg10()) {
+          throw new AssumptionViolatedException("postgresql >= 10 is needed ");
```
Comment (BillyYccc) [inline vertx-pg-client/src/test/java/io/vertx/pgclient/junit/ContainerPgRule.java]: can we use `Assume` in the testcase instead of checking the version here?

File: vertx-mysql-client/pom.xml (hunk 2 above)
Comment (benbenw) [reply vertx-mysql-client/pom.xml]: not needed will change

File: pom.xml (hunk 1 above)
Comment (benbenw) [reply pom.xml]: We should probably use it to set the testContainers version in mysql-client as well instead of moving it to the pg-client. What do you think ?

Comment (BillyYccc) [reply pom.xml]: ok but use the version which MySQL client is using now otherwise we may need a CQ for upgrading.

Comment (benbenw) [reply pom.xml]: What is a CQ ?

Comment (BillyYccc) [reply pom.xml]: https://wiki.eclipse.org/Development_Resources/Contribution_Questionnaire

File: vertx-pg-client/src/test/java/io/vertx/pgclient/junit/ContainerPgRule.java (hunk 3 above)
Comment (benbenw) [reply vertx-pg-client/src/test/java/io/vertx/pgclient/junit/ContainerPgRule.java]: I don't think so, the rule is loaded before the Assume in the test

Comment (BillyYccc) [reply vertx-pg-client/src/test/java/io/vertx/pgclient/junit/ContainerPgRule.java]: can you take https://github.com/eclipse-vertx/vertx-sql-client/blob/607e6ba04a250015be46d88917b8d11415bc3482/vertx-mysql-client/src/test/java/io/vertx/mysqlclient/MySQLUtilityCommandTest.java#L103 for reference?

Comment (benbenw) [reply vertx-pg-client/src/test/java/io/vertx/pgclient/junit/ContainerPgRule.java]: I have another PR (waiting for this one to be merged) which move the tests with embedded postgres to test container. 
Can this change be done later in the mentionned PR as it will require more change than just the `Assume` in the test.

**Comment by BillyYccc** [comment]
@vietj I think we need to file a CQ for the new dependency first.

**Comment by vietj** [comment]
can we test scram without testcontainer for now ?

I will make the CQ for both though

if we are going to use testcontainer for PG I would like rather a separate PR that migrates the project to use testcontainer for PG and a different one for scram.

**Comment by vietj** [comment]
actually we already have testcontainer CQ for postgresql

File: vertx-pg-client/pom.xml [hunk 4]
```diff
@@ -61,6 +61,13 @@
       <artifactId>vertx-sql-client</artifactId>
     </dependency>
 
+    <!-- sasl scram authentication -->
+    <dependency>
+      <groupId>com.ongres.scram</groupId>
```
Comment (vietj) [inline vertx-pg-client/pom.xml]: make it `<optional>true</optional>`

Comment (benbenw) [reply vertx-pg-client/pom.xml]: done

**Comment by benbenw** [comment]
@vietj 

> if we are going to use testcontainer for PG I would like rather a separate PR that migrates the project to use testcontainer for PG and a different one for scram.

I have a branch that migrates _most_ of the pg tests to test containers. I was waiting for this PR (scran) to be merged before pushing that branch.
If it's easier for you I can re-order my PRs to push the test migration first and then rebase this PR on top.

**Comment by vietj** [comment]
@benbenw can you make a separate PR for testcontainer that we merge first and then we finish review and merge this PR for scram ?

**Comment by benbenw** [comment]
> @benbenw can you make a separate PR for testcontainer that we merge first and then we finish review and merge this PR for scram ?

Yep ! Will do.

**Comment by benbenw** [comment]
I've created PR #431
This one will require a rebase as soon as the #431 is merged

**Comment by benbenw** [comment]
@vietj 
this PR was rebased

**Comment by vietj** [comment]
allright, we need to wait for the library CQ to be resolved.

**Comment by BillyYccc** [comment]
Indeed it would be good if you add something about the newly supported authentication method and the dependency used in the [documentation](https://github.com/eclipse-vertx/vertx-sql-client/blob/3.8/vertx-pg-client/src/main/asciidoc/index.adoc).

**Comment by benbenw** [comment]
minimalist doc added

**Comment by vietj** [comment]
sorry this has been pending for long time but we cannot yet use the scram library...

**Comment by benbenw** [comment]
@vietj no pb
the CQ for the scram lib is still pending or it has been refused ?

**Comment by vietj** [comment]
it is still pending

**Comment by vietj** [comment]
it seems actually resolved but it is not clear to me, you can see the CQ here https://dev.eclipse.org/ipzilla/process_bug.cgi . I asked about it in the comments.

**Comment by vietj** [comment]
this PR is good to go, would you main providing it for master instead ?

**Comment by vietj** [comment]
ping @benbenw

**Comment by benbenw** [comment]
Will rebase on master.
Give me a few days

**Comment by benbenw** [comment]
@vietj The PR was rebased.
Could you force a Travis build before merging ?

**Comment by vietj** [comment]
I can't find a way to do it @benbenw

**Comment by vietj** [comment]
I will checkout and test it myself

**Comment by vietj** [comment]
thanks for the contribution @benbenw
